tests/msc3083_test.go

```diff
@@ -127,5 +126,103 @@ func TestRestrictedRoomsLocalJoin(t *testing.T) {
 		},
 	)
 	// Fails since a fully invalid allow key rquires an invite.
-	FailJoinRoom(bob, t, room, "hs1")
+	FailJoinRoom(bob, t, room, "hs1", 403)
+}
+
+// Test joining a room with join rules restricted to membership in a space.
+func TestRestrictedRoomsRemoteJoin(t *testing.T) {
+	deployment := Deploy(t, "msc3083_remote", b.BlueprintFederationOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	// Create the space and put a room in it.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	space := alice.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+		"name":   "Space",
+	})
+	// The room is an unstable room version which supports the restricted join_rule.
+	room := alice.CreateRoom(t, map[string]interface{}{
+		"preset":       "public_chat",
+		"name":         "Room",
+		"room_version": "org.matrix.msc3083",
+		"initial_state": []map[string]interface{}{
+			{
+				"type":      "m.room.join_rules",
+				"state_key": "",
+				"content": map[string]interface{}{
+					"join_rule": "restricted",
+					"allow": []map[string]interface{}{
+						{
+							"space": &space,
+							"via":   []string{"hs1"},
+						},
+					},
+				},
+			},
+		},
+	})
+	alice.SendEventSynced(t, space, b.Event{
+		Type:     msc1772SpaceChildEventType,
+		StateKey: &room,
+		Content: map[string]interface{}{
+			"via": []string{"hs1"},
+		},
+	})
```

Comment:
is it worth factoring this out to a utility function?

Reply:
Yeah this part can, the parts with bob cannot since the HS differs between the two tests. :+1: